app/controllers/users/sessions_controller.rb

```diff
@@ -50,35 +50,31 @@ def destroy
       super
     end
 
-    def active
-      response.headers['Etag'] = '' # clear etags to prevent caching
-      session[:pinged_at] = now
-      Rails.logger.debug(alive?: alive?, expires_at: expires_at)
-      render json: { live: alive?, timeout: expires_at, remaining: remaining_session_time }
-    end
-
-    def keepalive
-      response.headers['Etag'] = '' # clear etags to prevent caching
-      session[:session_expires_at] = now + Devise.timeout_in if alive?
-      analytics.session_kept_alive if alive?
+    private
 
-      render json: { live: alive?, timeout: expires_at, remaining: remaining_session_time }
-    end
+    def redirect_with_flash_if_timeout
+      return if !params[:timeout]
+      case params[:timeout]
+      when 'session'
+        flash[:info] = t(
+          'notices.session_timedout',
+          app_name: APP_NAME,
+          minutes: IdentityConfig.store.session_timeout_in_minutes,
+        )
+      when 'form'
+        flash[:info] = t(
+          'notices.session_cleared',
+          minutes: IdentityConfig.store.session_timeout_in_minutes,
+        )
+      end
 
-    def timeout
-      analytics.session_timed_out
-      request_id = sp_session[:request_id]
-      sign_out
-      flash[:info] = t(
-        'notices.session_timedout',
-        app_name: APP_NAME,
-        minutes: IdentityConfig.store.session_timeout_in_minutes,
-      )
-      redirect_to root_url(request_id: request_id)
+      begin
+        redirect_to url_for(params.permit(:request_id))
+      rescue ActionController::UrlGenerationError # Binary data in parameters throw on redirect
+        head :bad_request
+      end
```

Comment:
since this is the last thing in the method, I'd take advantage of the method-level implicit rescue

```suggestion
      redirect_to url_for(params.permit(:request_id))
    rescue ActionController::UrlGenerationError # Binary data in parameters throw on redirect
      head :bad_request
```

Reply:
This is carried over from how [it's implemented today in `ApplicationController`](https://github.com/18F/identity-idp/blob/75b3b0e1a3d769125dceb8fb4c8c74a460648141/app/controllers/application_controller.rb#L166-L170). I assumed it was intentional to not make the `rescue` method-level to try to limit the scope of code that we're trying to rescue from, though I suppose that specific error probably wouldn't occur from anything else in the method aside from the `redirect_to`. Could still be useful for readability to understand what the `rescue` applies to?

Generally I'm in favor of flattening and using the method-level `rescue` though.

Reply:
In 1dd4971, I reverted a good chunk of this to the original handling, since this behavior doesn't just apply to the sign in screen.